controllers/global_search_setup.go

```diff
@@ -0,0 +1,570 @@
+				LastTransitionTime: metav1.Now(),
+			})
+			if updateErr != nil {
+				log.Error(updateErr, "Failed to update the Global Search status condition on Search CR instance.")
+			}
+		}
+	} else {
+		log.Info("The global-search-preview annotation is not present. Checking if global search was enabled before.")
+
+		// Use the status conditions to determine if global search was enabled before this reconcile.
+		globalSearchPresent := false
+		globalSearchConditionIndex := -1
+		for condIndex, condition := range instance.Status.Conditions {
+			if condition.Type == "GlobalSearchReady" {
+				globalSearchPresent = true
+				globalSearchConditionIndex = condIndex
+				break
+			}
+		}
+		if globalSearchPresent {
+			err := r.disableGlobalSearch(ctx, instance)
+			if err != nil {
+				log.Error(err, "Failed to disable global search.")
+				updateErr := r.updateGlobalSearchStatus(ctx, instance, metav1.Condition{
+					Type:               "GlobalSearchReady",
+					Status:             metav1.ConditionFalse,
+					Reason:             "GlobalSearchCleanupFailed",
+					Message:            "Failed to disable global search. " + err.Error(),
+					LastTransitionTime: metav1.Now(),
+				})
+				if updateErr != nil {
+					log.Error(updateErr, "Failed to update Search CR instance status.")
+				}
+				return &reconcile.Result{}, err
+			}
+
+			// Remove the status condition.
+			instance.Status.Conditions = append(instance.Status.Conditions[:globalSearchConditionIndex],
+				instance.Status.Conditions[globalSearchConditionIndex+1:]...)
+
+			err = r.commitSearchCRInstanceState(ctx, instance)
+			if err != nil {
+				log.Error(err, "Failed to update Search CR instance status.")
+				return &reconcile.Result{}, err
+			}
+		}
+	}
+	return &reconcile.Result{}, nil
+}
+
+// Validate pre-requisites.
+//
+//	a. MulticlusterGlobalHub operator is installed in the cluster.
+//	b. The ManagedServiceAccount add-on is enabled in the MultiClusterEngine CR.
+//	c. The ClusterProxy addon is enabled in the MultiClusterEngine CR.
+func (r *SearchReconciler) validateGlobalSearchPrerequisites(ctx context.Context) error {
+	log.V(5).Info("Checking global search pre-requisites.")
+	// Verify that MulticlusterGlobalHub operator is installed.
+	// oc get multiclusterglobalhub -A
+	multiclusterGlobalHub, err := r.DynamicClient.Resource(multiclusterGlobalHubGvr).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("MulticlusterGlobalHub operator is not installed.")
+	} else if len(multiclusterGlobalHub.Items) > 0 {
+		log.V(5).Info("Found MulticlusterGlobalHub intance.")
+	}
+
+	// Verify that the ManagedServiceAccount and ClusterProxy add-ons are enabled in the MultiClusterEngine CR.
+	mce, err := r.DynamicClient.Resource(multiclusterengineResourceGvr).
+		Get(ctx, "multiclusterengine", metav1.GetOptions{})
+	if err != nil {
+		log.Error(err, "Failed to get MulticlusterEngine CR.")
+		return fmt.Errorf("Failed to get MulticlusterEngine CR.")
```

Comment:
Attach original error message too?

Reply:
The original error message is logged. The message in the return error will get attached to the condition, so I prefer to keep it simple.